util/tls/tls_socket.cc

```diff
@@ -331,22 +331,254 @@ io::Result<TlsSocket::PushResult> TlsSocket::PushToEngine(const iovec* ptr, uint
+      break;
+    default:
+      LOG(DFATAL) << "Unsupported " << op_val;
+  }
+}
+
+void TlsSocket::AsyncWriteReq::Run() {
+  while (true) {
+    if (state == AsyncWriteReq::PushToEngine) {
+      // We never preempt here
+      io::Result<PushResult> push_res = owner->PushToEngine(vec, len);
+      if (!push_res) {
+        CompleteAsyncReq(make_unexpected(push_res.error()));
+        // We are done with this AsyncWriteReq. Caller might started
+        // a new one.
+        return;
+      }
+      last_push = *push_res;
+      state = AsyncWriteReq::HandleOpAsyncTag;
+    }
+
+    if (state == AsyncWriteReq::HandleOpAsyncTag) {
+      state = AsyncWriteReq::MaybeSendOutputAsyncTag;
+      if (last_push.engine_opcode < 0) {
+        HandleOpAsync(last_push.engine_opcode);
+        return;
+      }
+    }
+
+    if (state == AsyncWriteReq::MaybeSendOutputAsyncTag) {
+      state = AsyncWriteReq::PushToEngine;
+      if (last_push.written > 0) {
+        state = AsyncWriteReq::Done;
+        MaybeSendOutputAsync();
+        return;
+      }
+    }
+
+    if (state == AsyncWriteReq::Done) {
+      return CompleteAsyncReq(last_push.written);
+    }
+  }
+}
+
 void TlsSocket::AsyncWriteSome(const iovec* v, uint32_t len, io::AsyncProgressCb cb) {
-  io::Result<size_t> res = WriteSome(v, len);
-  cb(res);
+  CHECK(!async_write_req_.has_value());
+  async_write_req_.emplace(AsyncWriteReq(this, std::move(cb), v, len));
+  async_write_req_->Run();
+}
+
+void TlsSocket::AsyncReadReq::Run() {
+  DCHECK_GT(len, 0u);
+
+  while (true) {
+    DCHECK(!dest.empty());
+
+    size_t read_len = std::min(dest.size(), size_t(INT_MAX));
+
+    Engine::OpResult op_result = owner->engine_->Read(dest.data(), read_len);
+
+    int op_val = op_result;
+
+    DVLOG(2) << "Engine::Read " << dest.size() << " bytes, got " << op_val;
+
+    if (op_val > 0) {
+      read_total += op_val;
+
+      // I do not understand this code and what the hell I meant to do here. Seems to work
+      // though.
+      if (size_t(op_val) == read_len) {
+        if (size_t(op_val) < dest.size()) {
+          dest.remove_prefix(op_val);
+        } else {
+          ++vec;
+          --len;
+          if (len == 0) {
+            // We are done. Call completion callback.
+            CompleteAsyncReq(read_total);
+            return;
+          }
+          dest = Engine::MutableBuffer{reinterpret_cast<uint8_t*>(vec->iov_base), vec->iov_len};
+        }
+        // We read everything we asked for but there are still buffers left to fill.
+        continue;
+      }
+      break;
+    }
+
+    DCHECK(!continuation);
+    HandleOpAsync(op_val);
+    return;
+  }
+
+  // We are done. Call completion callback.
+  CompleteAsyncReq(read_total);
 }
 
-// TODO: to implement async functionality.
 void TlsSocket::AsyncReadSome(const iovec* v, uint32_t len, io::AsyncProgressCb cb) {
-  io::Result<size_t> res = ReadSome(v, len);
-  cb(res);
+  CHECK(!async_read_req_.has_value());
```

Comment:
This is not how I would design it. 
TlsEngine could already have a ready read buffer. you could return quickly by fetching it without doing any async stuff. Only if the buffer is empty it's worth dispatching the upstream read.

Reply:
Well, isn't that what I do here? We don't start/submit any async operation. We construct
the object and call `Run()` and within this function we will read the EngineBuffer first if there are bytes available -- so we are on the fast path already, that is, we don't dispatch upstream.
As I am changing it now to a unique ptr, we now always get an allocation. I think the cost would be insignificant because the allocator will be able to satisfy that without actually requesting more memory and this is not a common path anyways.